spec/ruby/core/module/define_method_spec.rb

```diff
@@ -499,6 +499,36 @@ def foo
       Class.new { define_method :bar, m }
     }.should raise_error(TypeError, /can't bind singleton method to a different class/)
   end
+
+  it "defines the new method public when the definition frame self is different from the target" do
+    foo_class = Class.new do
+      private def bar
+        "public"
+      end
+    end
+
+    foo = foo_class.new
+    foo.singleton_class.define_method(:bar, foo.method(:bar))
+
+    foo.bar.should == "public"
+  end
+
+  it "defines the new method according to the scope when the definition context is the same" do
+    FooWithOneBar = Class.new do
```

Comment:
Let's use a local variable here, I don't think this class needs to be named.
@andrykonchin Could you do that?